app/bundles/EmailBundle/Model/EmailModel.php

```diff
@@ -446,6 +448,39 @@ public function hitEmail($stat, $request, $viaBrowser = false)
         if ($this->dispatcher->hasListeners(EmailEvents::EMAIL_ON_OPEN)) {
             $event = new EmailOpenEvent($stat, $request, $firstTime);
             $this->dispatcher->dispatch(EmailEvents::EMAIL_ON_OPEN, $event);
+            //device granularity
+            $dd = new DeviceDetector($request->server->get('HTTP_USER_AGENT'));
+
+            $dd->parse();
+
+            $emailOpenDevice = new StatDevice();
+
+            $emailOpenDevice->setDateOpened($readDateTime->toUtcString());
+            $emailOpenDevice->setIpAddress($ipAddress);
+            $emailOpenDevice->setStat($stat->getId());
+            $emailOpenDevice->setChannel('Email');
+            $emailOpenDevice->setChannelId($email->getId());
```

Comment:
This needs to be something like `setChannelId((null !== $email) ? $email->getId() : null)` to not cause fatal errors for tracking opens for emails sent directly to contacts (through their profile/details page)